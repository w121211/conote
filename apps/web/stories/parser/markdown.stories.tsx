import React from 'react'
import { ComponentMeta, ComponentStory } from '@storybook/react'
import ReactMarkdown from 'react-markdown'
import { ReactMarkdownOptions } from 'react-markdown/lib/react-markdown'
import ParseRenderEl from '../../frontend/components/editor-textarea/src/components/inline/parse-render-el'

export default {
  title: 'parsers/ReactMarkdown',
  component: ReactMarkdown,
} as ComponentMeta<typeof ReactMarkdown>

// function rehypeWrapText(): Plugin {
//   return (tree: Root) => {
//     // visitParents(tree, 'text', (node, ancestors) => {
//     //   if (ancestors.at(-1).tagName !== 'custom-typography') {
//     //     node.type = 'element'
//     //     node.tagName = 'custom-typography'
//     //     node.children = [{ type: 'text', value: node.value }]
//     //   }
//     // })
//     visit(tree, 'text', (node, i, parent) => {
//       console.log(node, parent)
//     })
//   }
// }

export const Base = () => {
  const markdown = `A paragraph with *emphasis* and **strong importance**. [[hello world]]

#this is something cool-123456789#

#Which browser is best at keeping things confidential?-mock_discuss_0_active#

> A block quote with ~strikethrough~ and a URL: https://reactjs.org. [[hello world]]

* Lists
* [ ] todo
* [x] done

A table:

| a | b |
| - | - |
`

  const components: ReactMarkdownOptions['components'] = {
    p: ({ node, children, ...props }) => {
      return (
        <p {...props}>
          {children.map((e, i) => {
            if (typeof e === 'string') {
              return (
                <span key={i}>
                  <ParseRenderEl key={i} blockUid="blockUid" str={e} isViewer />
                </span>
              )
            } else return e
          })}
        </p>
      )
    },
  }

  return <ReactMarkdown components={components}>{markdown}</ReactMarkdown>
}
