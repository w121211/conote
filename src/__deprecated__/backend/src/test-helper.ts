import dotenv from 'dotenv'
import omitDeep from 'omit-deep-lodash'
import { removeUndefinedFields } from '../../lib/editor/src/test-helper'

export function _clean(obj: any): any {
  return removeUndefinedFields(omitDeep(obj, ['createdAt', 'updatedAt']))
}

const config = dotenv.config()
if (config.error) throw config.error
if (!config.parsed?.BOT_EMAIL || !config.parsed?.BOT_PASSWORD) {
  throw new Error('BOT_EMAIL or BOT_PASSWORD not found in .env')
}

export const BOT = { id: 'bot', email: config.parsed.BOT_EMAIL, password: config.parsed.BOT_PASSWORD }

export const TESTUSERS = [
  { id: 'test-user-1', email: 'aaa@aaa.com', password: 'aaa' },
  { id: 'test-user-2', email: 'bbb@bbb.com', password: 'bbb' },
  { id: 'test-user-3', email: 'ccc@ccc.com', password: 'ccc' },
  { id: 'test-user-4', email: 'ddd@ddd.com', password: 'ddd' },
  { id: 'test-user-5', email: 'eee@eee.com', password: 'eee' },
]

export const TESTOAUTHORS = [{ name: 'test-oauthor-1' }]
