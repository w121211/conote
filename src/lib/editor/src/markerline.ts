import { Markerheader, Markerline, ExtToken, Section } from './typing'
import { filterTokens, streamToStr, markerToStr, toStampMarkerlinesDict } from './helper'

/**
 * 將markerline插入至body-text中，插入的line會同時加入stamp
 * 注意：插入後line-number會變動，但這裡不更新line-number，需要另外叫`updateMarkerlines()`來修正
 * */
export function insertMarkerlinesToText(
  curMarkerheaders: Markerheader[],
  curMarkerlines: Markerline[],
  curText: string,
  insertMarkerlines: Markerline[],
  chance: { string: (a: any) => string },
): [string, Markerline[]] {
  // 將body分成一行一行，每行搭配對應的markerline, markerheader
  const lns: [string, { markerheader?: Markerheader; markerline?: Markerline }][] = curText
    .split('\n')
    .map<[string, { markerheader?: Markerheader; markerline?: Markerline }]>((e, i) => {
      const markerheader = curMarkerheaders.find(e => e.linenumber === i)
      const markerline = curMarkerlines.find(e => e.linenumber === i)
      return [e, { markerheader, markerline }]
    })

  function _insert(marekerline: Markerline, chance: { string: (a: any) => string }) {
    const a: Markerline = {
      // line-number需要之後再update
      linenumber: -1,
      str: marekerline.str,
      marker: marekerline.marker,
      new: true,
      src: marekerline.src,
      srcStamp: marekerline.stampId,
      oauthor: marekerline.oauthor,
      stampId: `%${chance.string({
        length: 3,
        pool: 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789',
      })}`,
    }

    if (a.marker) {
      const strWithStamp = `${markerToStr(a.marker, false)} ${a.stampId}`

      // 找對應的marker-header，並插入至header下一行
      const idx = lns.findIndex(e => e[1].markerheader?.marker.key && e[1].markerheader.marker.key === a.marker?.key)
      if (idx >= 0) {
        lns.splice(idx + 1, 0, [strWithStamp, { markerline: a }])
      } else {
        // 找不到marker-header的情況，插入一個header
        lns.push([a.marker.key, { markerheader: { linenumber: -1, marker: { key: a.marker.key } } }])
        lns.push([strWithStamp, { markerline: a }])
      }

      // const idx = findLastIndex(lns, e => e[1]?.marker?.mark === marekerline.marker?.mark)
    }
  }

  for (const e of insertMarkerlines) {
    if (e.marker && e.marker.key && e.marker.value) {
      _insert(e, chance)
    } else {
      console.error(`想插入的markerline不符合格式\n${e}`)
      // throw new Error('想插入的markerline不符合格式')
    }
  }

  // 將分行的body重新組合起來
  let joinedBody = ''
  const markerlines: Markerline[] = []
  for (const [str, { markerline }] of lns) {
    joinedBody += `${str}\n`
    if (markerline) {
      markerlines.push(markerline)
    }
  }

  return [joinedBody, markerlines]
}

/**
 * 和原本的markerlines比較，返回更新後的markerLines，過程中不會動到text
 * 1. 依照stamp更新markerline的line-number、value等等  TODO：需要check, validate
 * 2. 對新的line增加markerline
 */
export function updateMarkerlines(
  cur: Markerline[],
  sects: Section[],
  chance: { string: (a: any) => string },
  src?: string,
  oauthor?: string,
): {
  markerheaders: Markerheader[]
  markerlines: Markerline[]
} {
  // 建立、更新markerLines，同步檢查、建立stamp
  const dict = toStampMarkerlinesDict(cur)
  const markerlines: Markerline[] = []
  const markerheaders: Markerheader[] = []

  for (const sect of sects) {
    // 1. 針對marker-header，只需紀錄mark的linenumber
    for (const tk of filterTokens<ExtToken>(
      sect.stream ?? [],
      e => e.type === 'line-mark' || e.type === 'inline-mark',
    )) {
      if (tk.marker === undefined) {
        throw new Error('line-marker, inline-mark token一定要有marker')
      }
      markerheaders.push({
        linenumber: tk.linenumber,
        marker: tk.marker,
        inline: tk.type === 'inline-mark' ? true : undefined,
      })
    }

    // 2. 針對line value, inline-value
    for (const tk of filterTokens<ExtToken>(
      sect.stream ?? [],
      e => e.type === 'inline-value' || e.type === 'line-value',
    )) {
      if (tk.marker === undefined) {
        throw new Error('line-value, inline-value token一定要有marker')
      }

      const _updated = {
        linenumber: tk.linenumber,
        str: streamToStr(tk),
        marker: tk.marker,
        nestedCard: sect.nestedCard,
      }
      const stampTokens = filterTokens<ExtToken>(tk.content ?? [], e => e.type === 'stamp')

      if (stampTokens.length === 0) {
        // 沒有stamp，生成一個stmampId TODO: 要確保stampId不重複

        // 1. 辨識是否為comment, poll, reply, neat-reply，需為`[?]`
        let isComment: true | undefined
        let isPoll: true | undefined
        let isReply: true | undefined // TODO: 還未支援
        let isNeatReply: true | undefined
        let pollChoices

        if (tk.marker.key === '[?]') {
          const choices = filterTokens<ExtToken>(tk.content ?? [], e => e.type === 'vote-chocie')
          if (choices.length === 1) {
            isNeatReply = true
            pollChoices = choices.map<string>(e => streamToStr(e))
          } else if (choices.length > 0) {
            isPoll = true
            pollChoices = choices.map<string>(e => streamToStr(e))
          } else {
            isComment = true
          }
        }

        // 2. 生成stampId、存入
        markerlines.push({
          ..._updated,
          new: true,
          noStamp: true,
          src,
          oauthor,
          stampId: `%${chance.string({
            length: 3,
            pool: 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789',
          })}`,
          comment: isComment,
          poll: isPoll,
          pollChoices,
          neatReply: isNeatReply,
        })
        continue
      } else {
        // 有stamp TODO: validate

        // Hack: Prism-Parser不會在同一行內找到2個stamp-token，可直接用index=0來呼叫
        const _stamp = (stampTokens[0].content as string).trim()
        if (_stamp in dict) {
          // TODO: 和原始markerLine比較是否有更動
          markerlines.push({ ...dict[_stamp], ..._updated })
        } else {
          // 有anchor但broken
          // console.log(cur)
          // console.log(dict)
          // console.log({ ..._updated, stampId: _stamp, broken: true })
          markerlines.push({ ..._updated, stampId: _stamp, broken: true })
        }
      }
    }
  }

  // 依linenumber排序markerlines，不能省略，因為有些function會需要找最後一行
  markerheaders.sort((a, b) => a.linenumber - b.linenumber)
  markerlines.sort((a, b) => a.linenumber - b.linenumber)

  return { markerheaders, markerlines }
}
