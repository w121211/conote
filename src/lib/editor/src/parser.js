"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitByUrl = exports.findUrl = exports.tokenizeSection = exports.tokenizeSymbol = void 0;
const prismjs_1 = __importDefault(require("prismjs"));
const helper_1 = require("./helper");
const GRAMMAR = {
    'multiline-marker': {
        // pattern: /^\[[^\s\]]*\]$(?:\n^(?!\[).+)+/m,
        pattern: /^\[[^\s[\]]*\]$(?:\n^(?!\[[^\s[\]]*\]).+)+/m,
        inside: {
            'line-mark': {
                pattern: /^\[[^\s[\]]*\]$/m,
            },
            'line-value': {
                pattern: /^.+$/m,
                inside: {
                    ticker: { pattern: /\$[A-Z-]+/ },
                    topic: { pattern: /\[\[[^\]]+\]\]/u },
                    stamp: { pattern: /\s%[a-zA-Z0-9]{3}$/ },
                },
            },
            // 'list-string': {
            //     pattern: /^[-]+\s.+$/m,
            // },
        },
    },
    'inline-marker': {
        pattern: /^\[[^\s[\]]+\].*$/m,
        inside: {
            'inline-mark': {
                pattern: /^\[[^\s[\]]+\]/,
            },
            'inline-value': {
                pattern: /^.+$/,
                inside: {
                    ticker: { pattern: /\$[A-Z-]+/ },
                    topic: { pattern: /\[\[[^\]]+\]\]/u },
                    // TODO: 沒辦法將space與stamp分開（會成為一個string，需要trim）
                    stamp: { pattern: /\s%[a-zA-Z0-9]{3}$/ },
                },
            },
        },
    },
    ticker: { pattern: /\$[A-Z-]+/ },
    topic: { pattern: /\[\[[^\]]+\]\]/u },
    // 'radio': {
    //   pattern: /[\s\t]+\[\w?\]\p{L}+/u,
    //   // greedy: true
    // },
};
const SYMBOL_GRAMMAR = {
    ticker: GRAMMAR.ticker,
    topic: GRAMMAR.topic,
};
const SECTION_GRAMMAR = {
    'sect-ticker': {
        alias: 'sect-ticker',
        pattern: /^\n\$[A-Z-]+(@\w+)?$/m,
        inside: {
            'sect-symbol': { pattern: /^\n\$[A-Z-]+/ },
            'sect-user': { pattern: /@\w+/ },
        },
    },
    'sect-ticker-begin-line': {
        alias: 'sect-ticker',
        pattern: /^\$[A-Z-]+(@\w+)?\n/,
        inside: {
            'sect-symbol': { pattern: /^\$[A-Z-]+/ },
            'sect-user': { pattern: /@\w+/ },
        },
    },
    'sect-topic': {
        alias: 'sect-topic',
        pattern: /^\n\[\[[^\]]+\]\](@\w+)?$/m,
        inside: {
            'sect-symbol': { pattern: /^\n\[\[[^\]]+\]\]/u },
            'sect-user': { pattern: /@\w+/ },
        },
    },
    'sect-topic-begin-line': {
        alias: 'sect-topic',
        pattern: /^\[\[[^\]]+\]\](@\w+)?\n/,
        inside: {
            'sect-symbol': { pattern: /^\[\[[^\]]+\]\]/u },
            'sect-user': { pattern: /@\w+/ },
        },
    },
    'sect-breaker': {
        alias: 'sect-breaker',
        pattern: /^\n\/{3,}.+$/m,
    },
    'sect-url': {
        alias: 'sect-url',
        pattern: /^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&//=]*)$/m,
    },
};
function tokenizeSymbol(text) {
    /** 將text中的symbol($AA, [[Topic]])轉成token */
    return prismjs_1.default.tokenize(text, SYMBOL_GRAMMAR);
}
exports.tokenizeSymbol = tokenizeSymbol;
function validate(items, allowedMarkers) {
    for (const e of items) {
        const format = allowedMarkers.find(f => f.mark === e.mark);
        if (format === undefined) {
            e.error = '不在允許的markers中';
            continue;
        }
        if (e.value === undefined) {
            e.error = '沒有value';
            continue;
        }
        if (format.validater && !format.validater(e.value)) {
            e.error = 'value不符合格式';
            continue;
        }
        if (format.inline) {
            const filtered = items.filter(e => e.mark === format.mark);
            for (let i = 0; i < filtered.length - 1; i++) {
                filtered[i].error = '只能define一次，最後一個會被保留';
            }
        }
        // if (e.mark === MARKER_FORMAT.card.mark && getCardId) {
        //     getCardId(e.value)
        //         .then(result => {
        //             if (result === null) {
        //                 e.error = '找不到對應的card'
        //             } else {
        //                 e.cardId = result
        //             }
        //         })
        // }
        // if (e.children) {
        //     if (format.nested) {
        //         // TODO: 當前只有ticker可以nested，應該要靈活一些
        //         e.children = validate(e.children, TICKER_FORMATTER)
        //     } else {
        //         e.error = '不允許nested'
        //     }
        // }
    }
    return items;
}
function tokenizeSection(text, 
// symbolCardDict: Record<string, CardIdentifier>,
// nestedCards: CardIdentifier[] = [],
rootFormat, oauthorName, allowedSects = ['ticker', 'topic']) {
    // 分出text中的每個section
    // let _sect: {bodyTokens: []} = { bodyTokens: [] }
    var _a, _b;
    function _parseSectToken(token) {
        let symbol;
        let user;
        if (Array.isArray(token.content)) {
            for (const e of token.content) {
                if (typeof e === 'string') {
                    // do nothing
                }
                else if (e.type === 'sect-symbol') {
                    symbol = helper_1.streamToStr(e.content).trim();
                }
                else if (e.type === 'sect-user') {
                    user = helper_1.streamToStr(e.content);
                }
            }
            if (symbol === undefined) {
                console.error(token);
                throw new Error();
            }
            return { symbol, user };
        }
        console.error(token);
        throw new Error();
    }
    const sects = [];
    let _sect = {
        bodyTokens: [],
    };
    // let _sectToken: Prism.Token
    // let _bodyTokens: (Prism.Token | string)[] = []
    for (const e of prismjs_1.default.tokenize(text, SECTION_GRAMMAR)) {
        if (typeof e === 'string') {
            _sect.bodyTokens.push(e);
        }
        else if (e.alias === 'sect-ticker') {
            // 先儲存之前的section
            sects.push(Object.assign({}, _sect));
            // TODO: @ME
            const parsed = _parseSectToken(e);
            _sect = {
                ticker: true,
                nestedCard: { symbol: parsed.symbol, oauthor: (_a = parsed.user) !== null && _a !== void 0 ? _a : oauthorName },
                sectToken: e,
                bodyTokens: [],
            };
        }
        else if (e.alias === 'sect-topic') {
            sects.push(Object.assign({}, _sect));
            const parsed = _parseSectToken(e);
            _sect = {
                topic: true,
                nestedCard: { symbol: parsed.symbol, oauthor: (_b = parsed.user) !== null && _b !== void 0 ? _b : oauthorName },
                sectToken: e,
                bodyTokens: [],
            };
        }
        else if (e.alias === 'sect-breaker') {
            sects.push(Object.assign({}, _sect));
            // 把breaker視為獨立的section
            sects.push({ breaker: true, sectToken: e, bodyTokens: [] });
            // breaker之後的新section
            _sect = { bodyTokens: [] };
        }
        else {
            console.error(e);
            throw new Error('應該要處理但未處理的token-type');
        }
    }
    sects.push(Object.assign({}, _sect));
    for (const e of sects) {
        const body = helper_1.streamToStr(e.bodyTokens);
        if (e.ticker || e.topic) {
            // 對section做tokenize section-body
            e.bodyTokens = prismjs_1.default.tokenize(body, GRAMMAR);
        }
        else if (!e.breaker && body.length > 0) {
            // 所有非ticker、topic的card & 有body-string的section皆視為root
            e.root = true;
            e.bodyTokens = prismjs_1.default.tokenize(body, GRAMMAR);
        }
    }
    // 對每個token紀錄linenumber、marker
    let linenumber = 0;
    let mark = null;
    function _recursiveExtend(stream) {
        if (typeof stream === 'string') {
            linenumber += stream.split('\n').length - 1;
            return stream;
        }
        else if (Array.isArray(stream)) {
            return stream.map(e => _recursiveExtend(e));
        }
        else if (stream.type === 'inline-mark' || stream.type === 'line-mark') {
            mark = helper_1.streamToStr(stream.content);
            if (mark === '') {
                throw new Error();
            }
            return Object.assign(Object.assign({}, stream), { linenumber, marker: { mark }, content: _recursiveExtend(stream.content) });
        }
        else if (stream.type === 'inline-value' || stream.type === 'line-value') {
            const value = helper_1.streamToStr(stream.content, 'stamp').trim();
            if (mark === null) {
                throw new Error();
            }
            return Object.assign(Object.assign({}, stream), { linenumber, marker: { mark, value }, content: _recursiveExtend(stream.content) });
        }
        else {
            return Object.assign(Object.assign({}, stream), { linenumber, content: _recursiveExtend(stream.content) });
        }
    }
    for (const e of sects) {
        // section的全部stream（含section-token, body-tokens)
        let stream = [];
        if (e.sectToken)
            stream.push(e.sectToken);
        stream = stream.concat(e.bodyTokens);
        // 轉成ext-token
        mark = null;
        e.stream = _recursiveExtend(stream);
    }
    return sects.map(e => {
        return {
            root: e.root,
            breaker: e.breaker,
            ticker: e.ticker,
            topic: e.topic,
            nestedCard: e.nestedCard,
            stream: e.stream,
        };
    });
}
exports.tokenizeSection = tokenizeSection;
function findUrl(text) {
    /** 從首先出現的單行URL開始，擷取從URL以下的文章 */
    const tokens = prismjs_1.default.tokenize(text, SECTION_GRAMMAR);
    let url;
    const _tokensAfterUrl = [];
    for (const e of tokens) {
        if (url === undefined) {
            if (typeof e !== 'string' && e.type === 'sect-url')
                url = e.content;
            else
                continue;
        }
        _tokensAfterUrl.push(e);
    }
    // return { url, textAfterUrl: tokensToText(_tokensAfterUrl) }
    return { url, textAfterUrl: 'tokensToText(_tokensAfterUrl)' };
}
exports.findUrl = findUrl;
function splitByUrl(text) {
    /** 以text中的單行URL做split，返回:[url, part-text][] */
    const tokens = prismjs_1.default.tokenize(text, SECTION_GRAMMAR);
    let buffer = [];
    let url;
    const splits = [];
    for (const e of tokens) {
        if (typeof e !== 'string' && e.type === 'sect-url') {
            // 儲存前一個buffer後清空
            splits.push([url, helper_1.streamToStr(buffer)]);
            buffer = [];
            // 當前url
            url = e.content;
        }
        else {
            buffer.push(e);
        }
    }
    // 儲存最後一個
    splits.push([url, helper_1.streamToStr(buffer)]);
    return splits;
}
exports.splitByUrl = splitByUrl;
