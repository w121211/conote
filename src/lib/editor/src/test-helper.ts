import { readFileSync } from 'fs'
import { splitByUrl } from './parser'

export function load(filepath: string): [string, string][] {
  return (
    splitByUrl(readFileSync(filepath, { encoding: 'utf8' }))
      .filter((e): e is [string, string] => e[0] !== undefined)
      // .map(e => [e[0], `\n${e[1].trim()}`])
      .map(e => [e[0], e[1].trim()])
  )
}

export class FakeChance {
  i = 0
  string(): string {
    const str = this.i.toString(36).padStart(3, '0')
    this.i += 1
    if (this.i >= 36 * 36 * 36) {
      this.i = 0
    }
    return str
  }
}

export function removeUndefinedFields<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj))
}
