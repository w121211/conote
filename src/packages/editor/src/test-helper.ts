import { readFileSync } from 'fs'
import { splitByUrl } from './parser'

export function load(filepath: string): [string, string][] {
  return (
    splitByUrl(readFileSync(filepath, { encoding: 'utf8' }))
      .filter((e): e is [string, string] => e[0] !== undefined)
      // .map(e => [e[0], `\n${e[1].trim()}`])
      .map(e => [e[0], e[1].trim()])
  )
}

export class FakeChance {
  private i = 0

  public string(): string {
    const str = this.i.toString(36).padStart(3, '0')
    this.i += 1
    if (this.i >= 36 * 36 * 36) {
      this.i = 0
    }
    return str
  }

  public reset(): void {
    this.i = 0
  }
}

export function omitUndefined<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj))
}
