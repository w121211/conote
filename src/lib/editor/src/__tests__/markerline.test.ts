/* eslint-disable no-useless-escape */
import omitDeep from 'omit-deep-lodash'
import { tokenizeSection } from '../parser'
import { insertMarkerlinesToText, updateMarkerlines } from '../markerline'
import { load, FakeChance, removeUndefinedFields } from '../test-helper'

jest.mock('chance', () => {
  return {
    Chance: jest.fn().mockImplementation(() => {
      return {
        i: 0,
        string(): string {
          const str = this.i.toString(36).padStart(3, '0')
          this.i += 1
          if (this.i >= 36 * 36 * 36) {
            this.i = 0
          }
          return str
        },
      }
    }),
  }
})

const basicText = `[~] inline 111 inline 111
[=] inline 222 inline 222
[?]
<AAA> <BBB> poll 111
poll 222 <AAA> <BBB>
question 333?
- reply 111
- reply 222
[+]
point 111
point 222
[-]
[?] <CCC> neat reply
`

const simpleText = `[?] inline inline inline
[+]
111 111 %001
[-]
`

describe('updateMarkerlines fn', () => {
  const starter = updateMarkerlines(
    [],
    tokenizeSection(basicText),
    new FakeChance(),
    //   src?: string,
    // oauthor?: string,
  )
  const simpler = updateMarkerlines([], tokenizeSection(simpleText), new FakeChance())

  it('parse text to markerlines', () => {
    expect(removeUndefinedFields(starter.markerheaders)).toMatchSnapshot()
    expect(removeUndefinedFields(starter.markerlines)).toMatchSnapshot()
  })

  it('parse swaped lines', () => {
    const t1 = `[~]
[+]
111 111 %001
[?] inline inline inline %000
[-]`
    const parsed1 = updateMarkerlines(simpler.markerlines, tokenizeSection(t1), new FakeChance())
    expect(parsed1.markerlines).toMatchSnapshot()
  })

  it('parse new added lines', () => {
    const t1 = `[~]
[?] inline inline inline %000
[+]
111 111 %001
222 222
[-]
333 333`
    const parsed1 = updateMarkerlines(simpler.markerlines, tokenizeSection(t1), new FakeChance())
    expect(parsed1.markerlines).toMatchSnapshot()
  })
})

describe('insertMarkerlinesToBody fn', () => {
  const starterBody = `[~] 111 222 333
[?]
<AAA> <BBB> question quesiton
[+]
111 111
[-]
[多了空格1]\s
[多了空格2]\s
多了空格2.1 多了空格2.1 多了空格2.1
多了空格2.2 多了空格2.2 多了空格2.2
`
  const starter = updateMarkerlines([], tokenizeSection(starterBody), new FakeChance())

  it('insert markerline to existed marker-header', () => {
    const [body, markerlines] = insertMarkerlinesToText(
      starter.markerheaders,
      starter.markerlines,
      starterBody,
      [
        {
          linenumber: 10,
          str: 'insert insert insert',
          marker: {
            key: '[+]',
            value: 'insert insert insert',
          },
        },
      ],
      new FakeChance(),
    )
    expect(body).toMatchSnapshot()
    expect(markerlines).toMatchSnapshot()
  })

  // it('insert markerline to non existed marker-header', () => {
  //   //   const cur: Markerline[] = [
  //   //   ]
  //   // body: string,
  //   // insert: Markerline[],
  //   // chance: { string: (a: any) => string },
  //   //   insertMarkerlinesToBody()
  // })

  // it('insert inline markerline', () => {
  //   //   const cur: Markerline[] = [
  //   //   ]
  //   // body: string,
  //   // insert: Markerline[],
  //   // chance: { string: (a: any) => string },
  //   //   insertMarkerlinesToBody()
  // })
})
