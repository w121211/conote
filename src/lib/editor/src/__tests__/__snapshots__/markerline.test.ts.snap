// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insertMarkerlinesToBody fn insert markerline to existed marker-header 1`] = `
"[~] 111 222 333
[?]
<AAA> <BBB> question quesiton
[+]
insert insert insert %000
111 111
[-]
[多了空格1]s
[多了空格2]s
多了空格2.1 多了空格2.1 多了空格2.1
多了空格2.2 多了空格2.2 多了空格2.2

"
`;

exports[`insertMarkerlinesToBody fn insert markerline to existed marker-header 2`] = `
Array [
  Object {
    "comment": undefined,
    "linenumber": 0,
    "marker": Object {
      "key": "[~]",
      "value": "111 222 333",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": undefined,
    "pollChoices": undefined,
    "src": undefined,
    "stampId": "%000",
    "str": " 111 222 333",
  },
  Object {
    "comment": undefined,
    "linenumber": 2,
    "marker": Object {
      "key": "[?]",
      "value": "<AAA> <BBB> question quesiton",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": true,
    "pollChoices": Array [
      "<AAA>",
      "<BBB>",
    ],
    "src": undefined,
    "stampId": "%001",
    "str": "<AAA> <BBB> question quesiton",
  },
  Object {
    "linenumber": -1,
    "marker": Object {
      "key": "[+]",
      "value": "insert insert insert",
    },
    "new": true,
    "oauthor": undefined,
    "src": undefined,
    "srcStamp": undefined,
    "stampId": "%000",
    "str": "insert insert insert",
  },
  Object {
    "comment": undefined,
    "linenumber": 4,
    "marker": Object {
      "key": "[+]",
      "value": "111 111",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": undefined,
    "pollChoices": undefined,
    "src": undefined,
    "stampId": "%002",
    "str": "111 111",
  },
  Object {
    "comment": undefined,
    "linenumber": 6,
    "marker": Object {
      "key": "[多了空格1]",
      "value": "s",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": undefined,
    "pollChoices": undefined,
    "src": undefined,
    "stampId": "%003",
    "str": "s",
  },
  Object {
    "comment": undefined,
    "linenumber": 7,
    "marker": Object {
      "key": "[多了空格2]",
      "value": "s",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": undefined,
    "pollChoices": undefined,
    "src": undefined,
    "stampId": "%004",
    "str": "s",
  },
]
`;

exports[`updateMarkerlines fn parse new added lines 1`] = `
Array [
  Object {
    "comment": true,
    "linenumber": 1,
    "marker": Object {
      "key": "[?]",
      "value": "inline inline inline",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": undefined,
    "pollChoices": undefined,
    "src": undefined,
    "stampId": "%000",
    "str": " inline inline inline %000",
  },
  Object {
    "broken": true,
    "linenumber": 3,
    "marker": Object {
      "key": "[+]",
      "value": "111 111",
    },
    "nestedCard": undefined,
    "stampId": "%001",
    "str": "111 111 %001",
  },
  Object {
    "comment": undefined,
    "linenumber": 4,
    "marker": Object {
      "key": "[+]",
      "value": "222 222",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": undefined,
    "pollChoices": undefined,
    "src": undefined,
    "stampId": "%000",
    "str": "222 222",
  },
  Object {
    "comment": undefined,
    "linenumber": 6,
    "marker": Object {
      "key": "[-]",
      "value": "333 333",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": undefined,
    "pollChoices": undefined,
    "src": undefined,
    "stampId": "%001",
    "str": "333 333",
  },
]
`;

exports[`updateMarkerlines fn parse swaped lines 1`] = `
Array [
  Object {
    "broken": true,
    "linenumber": 2,
    "marker": Object {
      "key": "[+]",
      "value": "111 111",
    },
    "nestedCard": undefined,
    "stampId": "%001",
    "str": "111 111 %001",
  },
  Object {
    "comment": true,
    "linenumber": 3,
    "marker": Object {
      "key": "[?]",
      "value": "inline inline inline",
    },
    "neatReply": undefined,
    "nestedCard": undefined,
    "new": true,
    "noStamp": true,
    "oauthor": undefined,
    "poll": undefined,
    "pollChoices": undefined,
    "src": undefined,
    "stampId": "%000",
    "str": " inline inline inline %000",
  },
]
`;

exports[`updateMarkerlines fn parse text to markerlines 1`] = `
Array [
  Object {
    "inline": true,
    "linenumber": 0,
    "marker": Object {
      "key": "[~]",
    },
  },
  Object {
    "inline": true,
    "linenumber": 1,
    "marker": Object {
      "key": "[=]",
    },
  },
  Object {
    "linenumber": 2,
    "marker": Object {
      "key": "[?]",
    },
  },
  Object {
    "linenumber": 8,
    "marker": Object {
      "key": "[+]",
    },
  },
  Object {
    "inline": true,
    "linenumber": 11,
    "marker": Object {
      "key": "[-]",
    },
  },
  Object {
    "inline": true,
    "linenumber": 12,
    "marker": Object {
      "key": "[?]",
    },
  },
]
`;

exports[`updateMarkerlines fn parse text to markerlines 2`] = `
Array [
  Object {
    "linenumber": 0,
    "marker": Object {
      "key": "[~]",
      "value": "inline 111 inline 111",
    },
    "new": true,
    "noStamp": true,
    "stampId": "%000",
    "str": " inline 111 inline 111",
  },
  Object {
    "linenumber": 1,
    "marker": Object {
      "key": "[=]",
      "value": "inline 222 inline 222",
    },
    "new": true,
    "noStamp": true,
    "stampId": "%001",
    "str": " inline 222 inline 222",
  },
  Object {
    "linenumber": 3,
    "marker": Object {
      "key": "[?]",
      "value": "<AAA> <BBB> poll 111",
    },
    "new": true,
    "noStamp": true,
    "poll": true,
    "pollChoices": Array [
      "<AAA>",
      "<BBB>",
    ],
    "stampId": "%002",
    "str": "<AAA> <BBB> poll 111",
  },
  Object {
    "linenumber": 4,
    "marker": Object {
      "key": "[?]",
      "value": "poll 222 <AAA> <BBB>",
    },
    "new": true,
    "noStamp": true,
    "poll": true,
    "pollChoices": Array [
      "<AAA>",
      "<BBB>",
    ],
    "stampId": "%003",
    "str": "poll 222 <AAA> <BBB>",
  },
  Object {
    "linenumber": 9,
    "marker": Object {
      "key": "[+]",
      "value": "point 111",
    },
    "new": true,
    "noStamp": true,
    "stampId": "%004",
    "str": "point 111",
  },
  Object {
    "linenumber": 10,
    "marker": Object {
      "key": "[+]",
      "value": "point 222",
    },
    "new": true,
    "noStamp": true,
    "stampId": "%005",
    "str": "point 222",
  },
  Object {
    "linenumber": 12,
    "marker": Object {
      "key": "[?]",
      "value": "<CCC> neat reply",
    },
    "neatReply": true,
    "new": true,
    "noStamp": true,
    "pollChoices": Array [
      "<CCC>",
    ],
    "stampId": "%006",
    "str": " <CCC> neat reply",
  },
]
`;
